src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,202 @@
+            } else if (validAction && !validIndex) {
+                throw new DukeException("☹ OOPS!!! Please enter a valid task number to mark/unmark");
+            }
+            return null;
+        } catch (NumberFormatException e) {
+            return null;
+        }
+    }
+
+    /**
+     * This method takes in the string input of the user
+     * and return the task from the taskList to be added.
+     * @param str This is the user's string input
+     * @return Task The task to be added to the user's task list
+     * @throws DukeException This tells the user whether the task is missing a date or description
+     */
+    private static Task getTaskToAdd(String str) throws DukeException {
+        String[] splitStr = str.split(" ");
+        String type = splitStr[0];
+        if (type.equals("todo")) {
+            if (splitStr.length < 2) {
+                throw new DukeException("☹ OOPS!!! The description of a todo cannot be empty.\n");
+            }
+            String description = str.substring(type.length() + 1);
+            return new ToDo(description);
+        } else if (type.equals("deadline")) {
+            if (splitStr.length < 2) {
+                throw new DukeException("☹ OOPS!!! The description of a deadline cannot be empty.\n");
+            }
+            if (str.indexOf("/by") - 1 < 0) {
+                throw new DukeException("☹ OOPS!!! Please set date of deadline with /by.\n");
+            }
+            String description = str.substring(type.length() + 1, str.indexOf("/by") - 1);
+            String date = str.substring(str.indexOf("/by") + 4);
+            String formattedDate = Parser.parseDate(date);
+            return new Deadline(description, formattedDate);
+        } else if (type.equals("event")) {
+            if (splitStr.length < 2) {
+                throw new DukeException("☹ OOPS!!! The description of an event cannot be empty.\n");
+            }
+            if (str.indexOf("/at") - 1 < 0) {
+                throw new DukeException("☹ OOPS!!! Please set date of event with /at.\n");
+            }
+            String description = str.substring(type.length() + 1, str.indexOf("/at") - 1);
+            String date = str.substring(str.indexOf("/at") + 4);
+            String formattedDate = Parser.parseDate(date);
+            return new Event(description, formattedDate);
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * This method takes in the users string input
+     * and returns a list of tasks that contains
+     * the user's search keyword
+     * @param str Command input from user
+     * @param taskList User's complete task list
+     * @return list of tasks containing user's search keyword
+     * @throws DukeException This tells the user if no such task contains the user's search keyword
+     */
+    private static List<Task> getAllTaskWithKeyword(String str, TaskList taskList) throws DukeException {
+        String[] splitStr = str.split(" ");

```

Comment:
There's a bit of code duplication here where the idea of "splitting and matching the first word, then getting the rest of the string" is repeated multiple times. Can this common idea be perhaps abstracted away, perhaps with other functions like this?
```java
public static boolean isOfTheCommand(String input, String commandName) {
    String[] splitStr = input.split(" ");
    return splitStr[0].equals(commandName));
}
public static boolean getCommandBody(String input) {
    String[] splitStr = input.split(" ");
    String type = splitStr[0];
    return input.substring(type.length() + 1);
}
```